client/client.go

```diff
@@ -388,19 +390,87 @@ func createClientWithKeyspace(
 	return c, nil
 }
 
-// NewClientWithKeyspaceName creates a client with context and the specified keyspace name.
-func NewClientWithKeyspaceName(
+// APIVersion is the API version the server and the client is using.
+// See more details in https://github.com/tikv/rfcs/blob/master/text/0069-api-v2.md#kvproto
+type APIVersion int
+
+// The API versions the client supports.
+const (
+	V1 APIVersion = iota
+	V1TTL
+	V2
+)
+
+// APIContext is the context for API version.
+type APIContext interface {
+	GetAPIVersionInfo() (apiVersion APIVersion, keyspaceName string)
+}
+
+type apiContextV1 struct{}
+
+// NewAPIContextV1 creates a API context for V1.
+func NewAPIContextV1() APIContext {
+	return &apiContextV1{}
+}
+
+// GetAPIVersionInfo returns the API version and the keyspace name.
+func (apiCtx *apiContextV1) GetAPIVersionInfo() (version APIVersion, keyspaceName string) {
+	return V1, ""
+}
+
+type apiContextV1TTL struct{}
+
+// NewAPIContextV1TTL creates a API context for V1TTL.
+func NewAPIContextV1TTL() APIContext {
+	return &apiContextV1TTL{}
+}
+
+// GetAPIVersionInfo returns the API version and the keyspace name.
+func (apiCtx *apiContextV1TTL) GetAPIVersionInfo() (version APIVersion, keyspaceName string) {
+	return V1TTL, ""
+}
+
+type apiContextV2 struct {
+	keyspaceName string
+}
+
+// NewAPIContextV2 creates a API context with the specified keyspace name for V2.
+func NewAPIContextV2(keyspaceName string) APIContext {
+	return &apiContextV2{keyspaceName: keyspaceName}
```

Comment:
How about moving the deriving logic here:
```go
if keyspaceName == "" {
	keyspaceName = "DEFAULT"
}
return &apiContextV2{keyspaceName: keyspaceName}
```

Reply:
+1